tests/plugins/test_change_from_in_df.py

```diff
@@ -14,80 +14,171 @@
+def mock_workflow(df_path, parent_images=None, image_id_map=None):
     if MOCK:
         mock_docker()
 
+    parent_images = parent_images or {BASE_IMAGE: BASE_IMAGE_UNIQUE}
+    image_id_map = image_id_map or {BASE_IMAGE_UNIQUE: BASE_IMAGE_ID}
     builder = flexmock(
         image_id='test-builder-image',
         df_path=df_path,
-        base_image=ImageName(repo='spam', tag='eggs'),
+        base_image=ImageName.parse(BASE_IMAGE),
+        parent_images=parent_images,
         source=flexmock(
             path=os.path.dirname(str(df_path)),
             dockerfile_path=str(df_path),
         ),
+        tasker=flexmock(
+            inspect_image=lambda img: dict(Id=image_id_map[img])
+        ),
+        set_base_image=lambda img: setattr(builder, 'base_image', ImageName.parse(img)),
     )
 
     workflow = DockerBuildWorkflow(SOURCE, "test-image")
     workflow.builder = builder
-    workflow._base_image_inspect = inspect_data
 
     return workflow
 
 
-def runner(docker_tasker, workflow):
-    return PreBuildPluginsRunner(
-       docker_tasker, workflow,
+def run_plugin(workflow, allow_failure=False):
+    result = PreBuildPluginsRunner(
+       docker_tasker(), workflow,
        [{
           'name': ChangeFromPlugin.key,
           'args': {},
        }]
-    )
-    runner.run()
+    ).run()
+
+    if not allow_failure:  # exceptions are captured in plugin result
+        assert result[ChangeFromPlugin.key] is None, "Plugin threw exception, check logs"
+
+    return result[ChangeFromPlugin.key]
 
 
-def test_update_base_image(docker_tasker, tmpdir):  # noqa
+def test_update_base_image(tmpdir):
     df_content = dedent("""\
-        FROM fedora
+        FROM {}
         LABEL horses=coconuts
         CMD whoah
     """)
     dfp = df_parser(str(tmpdir))
-    dfp.content = df_content
+    dfp.content = df_content.format(BASE_IMAGE)
 
-    workflow = mock_workflow(dfp.dockerfile_path, dict(Id='bacon'))
-    runner(docker_tasker, workflow).run()
+    workflow = mock_workflow(dfp.dockerfile_path)
+    run_plugin(workflow)
 
-    expected_df = dedent("""\
-        FROM bacon
-        LABEL horses=coconuts
-        CMD whoah
-    """)
+    expected_df = df_content.format(BASE_IMAGE_ID)
     assert dfp.content == expected_df
 
 
-def test_update_base_image_broken(docker_tasker, tmpdir):  # noqa
-    df_content = " # nothing "
+def test_update_base_image_inspect_broken(tmpdir, caplog):
+    # exercise code branch where the base image inspect comes back without an Id
+    df_content = "FROM " + BASE_IMAGE
     dfp = df_parser(str(tmpdir))
     dfp.content = df_content
 
-    workflow = mock_workflow(dfp.dockerfile_path, dict(no_id="here"))
-    result = runner(docker_tasker, workflow).run()
+    workflow = mock_workflow(dfp.dockerfile_path)
+    workflow.builder.tasker.inspect_image = lambda img: dict(no_id="here")
+
+    with pytest.raises(NoIdInspection):
+        ChangeFromPlugin(docker_tasker(), workflow).run()
```

Comment:
run_plugin could take a second optional argument of "should_raise" that defaults to None and do pytest.raises(should_raise). I suggest that because after nicely encapsulating ChangeFromPlugin in your run_plugin() function, it's jarring to see it again being called directly in the tests cases.

Reply:
There are a number of situations where it makes more sense to go straight to the class to test a particular behavior, and there's a very specific reason for it here.

I like to raise named exceptions so I can test for exactly the thing that is expected to go wrong without relying on the text. But if you import the exception class from the plugin, and then run the plugin via a plugin-runner, the comparison of the exception you got to the expected class will fail. This is because the plugin runner reloads all the modules before running the plugin and what you get back is technically a different class, even though it came from the same code. So the level of indirection really gets in the way here.